<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="metasnf">
<title>A Simple Example • metasnf</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="A Simple Example">
<meta property="og:description" content="metasnf">
<meta name="google-site-verification" content="Xzz1SIDqvPwru_mXKCXAkX2pNfOz5PO_QO02QG98tZI">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">metasnf</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/getting_started.html">Getting Started</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Example Pipelines</h6>
    <a class="dropdown-item" href="../articles/a_simple_example.html">A Simple Example</a>
    <a class="dropdown-item" href="../articles/a_less_simple_example.html">A Less Simple Example</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Essential Objects</h6>
    <a class="dropdown-item" href="../articles/settings_matrix.html">The Settings Matrix</a>
    <a class="dropdown-item" href="../articles/data_list.html">The Data List</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Advanced Functionality</h6>
    <a class="dropdown-item" href="../articles/distance_metrics.html">Distance Metrics</a>
    <a class="dropdown-item" href="../articles/clustering_algorithms.html">Clustering Algorithms</a>
    <a class="dropdown-item" href="../articles/snf_schemes.html">SNF schemes</a>
    <a class="dropdown-item" href="../articles/feature_weights.html">Feature Weights</a>
    <a class="dropdown-item" href="../articles/stability_measures.html">Stability Measures and Consensus Clustering</a>
    <a class="dropdown-item" href="../articles/confounders.html">Confounders and Unwanted Signal</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Additional Plotting</h6>
    <a class="dropdown-item" href="../articles/similarity_matrix_heatmap.html">Similarity Matrix Heatmaps</a>
    <a class="dropdown-item" href="../articles/manhattan_plots.html">Manhattan Plots</a>
    <a class="dropdown-item" href="../articles/alluvial_plots.html">Alluvial Plots</a>
    <a class="dropdown-item" href="../articles/correlation_plots.html">Correlation Plots</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>A Simple Example</h1>
            
      
      
      <div class="d-none name"><code>a_simple_example.Rmd</code></div>
    </div>

    
    
<style>
div.aside { background-color:#fff2e6; }
</style>
<p>Download a copy of the vignette to follow along here: <a href="https://raw.githubusercontent.com/BRANCHlab/metasnf/main/vignettes/a_simple_example.Rmd" class="external-link">a_simple_example.Rmd</a></p>
<p>In this vignette, we will show how <em>r-package</em> can be used for
a very simple SNF workflow.</p>
<p>This simple workflow is the example of SNF provided in the original
<em>SNFtool</em> package. You can find the example by loading the
<em>SNFtool</em> package and then viewing the documentation for the main
SNF function by running <code><a href="https://rdrr.io/pkg/SNFtool/man/SNF.html" class="external-link">?SNF</a></code>.</p>
<div class="section level2">
<h2 id="the-original-snf-example">The original SNF example<a class="anchor" aria-label="anchor" href="#the-original-snf-example"></a>
</h2>
<div class="section level3">
<h3 id="load-the-package">1. Load the package<a class="anchor" aria-label="anchor" href="#load-the-package"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SNFtool</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="set-snf-hyperparameters">2. Set SNF hyperparameters<a class="anchor" aria-label="anchor" href="#set-snf-hyperparameters"></a>
</h3>
<p>Three hyperparameters are introduced in this example: <em>K</em>,
<em>alpha</em> (also referred to as sigma or eta in different
documentations), and <em>T</em>. You can learn more about the
significance of these hyperparameters in the original SNF paper (see
references).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="cn">T</span> <span class="op">&lt;-</span> <span class="fl">20</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-the-data">3. Load the data<a class="anchor" aria-label="anchor" href="#load-the-data"></a>
</h3>
<p>The SNFtool package provides two mock dataframes titled
<em>Data1</em> and <em>Data2</em> for this example. <em>Data1</em>
contains gene expression values of two genes for 200 patients.
<em>Data2</em> similarly contains methylation data for two genes for
those same 200 patients.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Data1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Data2</span><span class="op">)</span></span></code></pre></div>
<p>Here’s what the mock data looks like:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"heatmap.png"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># gene expression data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span></span>
<span>    <span class="va">Data1</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    show_rownames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    show_colnames <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># methylation data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span></span>
<span>    <span class="va">Data2</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    show_rownames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    show_colnames <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The “ground truth” of how this data was generated was that patients 1
to 100 were drawn from one distribution and patients 101 to 200 were
drawn from another. We don’t have access to that kind of knowledge in
real data, but we do here.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">true_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">100</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="generate-similarity-matrices-for-each-data-source">4. Generate similarity matrices for each data source<a class="anchor" aria-label="anchor" href="#generate-similarity-matrices-for-each-data-source"></a>
</h3>
<p>We consider the two gene expression variables in <em>Data1</em> to
contain information from one broader gene expression source and the two
methylation variables in <em>Data2</em> to contain information from a
broader methylation source.</p>
<p>The next step is to determine, <strong>for each of the sources we
have</strong>, how similar all of our patients are to each other.</p>
<p>This is done by first determining how <em>dissimilar</em> the
patients are to each other for each source, and then converting that
dissimilarity information into similarity information.</p>
<p>To calculate dissimilarity, we’ll use Euclidean distance.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">distance_matrix_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="va">Data1</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">distance_matrix_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="va">Data2</span>, method <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Then, we can use the <code>affinityMatrix</code> function provided by
<em>SNFtool</em> to convert those distance matrices into similarity
matrices.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">similarity_matrix_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNFtool/man/affinityMatrix.html" class="external-link">affinityMatrix</a></span><span class="op">(</span><span class="va">distance_matrix_1</span>, <span class="va">K</span>, <span class="va">alpha</span><span class="op">)</span></span>
<span><span class="va">similarity_matrix_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNFtool/man/affinityMatrix.html" class="external-link">affinityMatrix</a></span><span class="op">(</span><span class="va">distance_matrix_2</span>, <span class="va">K</span>, <span class="va">alpha</span><span class="op">)</span></span></code></pre></div>
<p>Those similarity matrices can be passed into the <code>SNF</code>
function to integrate them into a single similarity matrix that
describes how similar the patients are to each other across both the
gene expression and methylation data.</p>
</div>
<div class="section level3">
<h3 id="integrate-similarity-matrices-with-snf">5. Integrate similarity matrices with SNF<a class="anchor" aria-label="anchor" href="#integrate-similarity-matrices-with-snf"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fused_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNFtool/man/SNF.html" class="external-link">SNF</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">similarity_matrix_1</span>, <span class="va">similarity_matrix_2</span><span class="op">)</span>,</span>
<span>    <span class="va">K</span>,</span>
<span>    <span class="cn">T</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="find-clusters-in-the-integrated-matrix">6. Find clusters in the integrated matrix<a class="anchor" aria-label="anchor" href="#find-clusters-in-the-integrated-matrix"></a>
</h3>
<p>If we think there are 2 clusters in the data, we can use spectral
clustering to find 2 clusters in the fused network.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">number_of_clusters</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="va">assigned_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SNFtool/man/spectralClustering.html" class="external-link">spectralClustering</a></span><span class="op">(</span><span class="va">fused_network</span>, <span class="va">number_of_clusters</span><span class="op">)</span></span></code></pre></div>
<p>Sure enough, we are able to obtain the correct cluster label for all
patients.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">true_label</span> <span class="op">==</span> <span class="va">assigned_clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="the-same-example-using-r-package">The same example using <em>r-package</em><a class="anchor" aria-label="anchor" href="#the-same-example-using-r-package"></a>
</h2>
<p>The purpose of <em>r-package</em> is primarily to aid users explore a
wide possible range of solutions. Recreating the example provided with
the original <code>SNF</code> function will be an extremely restricted
usage of the package, but will reveal, broadly, how <em>r-package</em>
works.</p>
<div class="section level3">
<h3 id="load-the-package-1">1. Load the package<a class="anchor" aria-label="anchor" href="#load-the-package-1"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://branchlab.github.io/metasnf/">metasnf</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="store-the-data-in-a-data_list">2. Store the data in a <em>data_list</em><a class="anchor" aria-label="anchor" href="#store-the-data-in-a-data_list"></a>
</h3>
<p>All the data we’re working with will get stored in a single object
called the <code>data_list</code>. The <code>data_list</code> is made by
passing in each dataframe into the <code>generate_data_list</code>
function, alongside information about the name of the dataframe, the
broader source (referred to in this package as a “domain”) of
information that dataframe comes from, and the type of variables that
are stored inside that dataframe (can be continuous, discrete, ordinal,
categorical, or mixed).</p>
<p>The <code>data_list</code> generation process also requires you to
specify which column contains information about the ID of the patients.
In this case, that information isn’t there, so we’ll have to add it
ourselves. The added IDs span from 101 onwards (rather than from 1
onwards) purely for convenience: automatic sorting of patient names
won’t result in patient 199 being placed before patient 2.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add "patient_id" column to each dataframe</span></span>
<span><span class="va">Data1</span><span class="op">$</span><span class="st">"patient_id"</span> <span class="op">&lt;-</span> <span class="fl">101</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Data1</span><span class="op">)</span> <span class="op">+</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">Data2</span><span class="op">$</span><span class="st">"patient_id"</span> <span class="op">&lt;-</span> <span class="fl">101</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">Data2</span><span class="op">)</span> <span class="op">+</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_data_list.html">generate_data_list</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">Data1</span>,</span>
<span>        name <span class="op">=</span> <span class="st">"genes_1_and_2_exp"</span>,</span>
<span>        domain <span class="op">=</span> <span class="st">"gene_expression"</span>,</span>
<span>        type <span class="op">=</span> <span class="st">"continuous"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">Data2</span>,</span>
<span>        name <span class="op">=</span> <span class="st">"genes_1_and_2_meth"</span>,</span>
<span>        domain <span class="op">=</span> <span class="st">"gene_methylation"</span>,</span>
<span>        type <span class="op">=</span> <span class="st">"continuous"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    uid <span class="op">=</span> <span class="st">"patient_id"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The first entries are all lists which contains the following
elements:</p>
<ol style="list-style-type: decimal">
<li>The actual dataframe</li>
<li>A name for the dataframe (string)</li>
<li>A name for the <em>domain</em> of information the dataframe is
representative of (string)</li>
<li>The type of variable stored in the dataframe (options are
continuous, discrete, ordinal, categorical, and mixed)</li>
</ol>
<p>Finally, there’s an argument for the <code>uid</code> (the column
name that currently uniquely identifies all the subjects in your
data).</p>
<p>Behind the scenes, this function is building a nested list that keeps
track of all this information, but it is also:</p>
<ul>
<li>Converting the UID of the data into “subjectkey”</li>
<li>Removing all observations that contain any NAs</li>
<li>Removing all subjects who are not present in all input
dataframes</li>
<li>Arranging the subjects in all the dataframe by their UID</li>
<li>Prefixing the UID values with the string “subject_” to help with
cluster result characterization</li>
</ul>
<p>Any rows containing NAs are removed. If you don’t want a bunch of
your data to get removed because there are a few NAs sprinkled around
here and there, consider using <a href="https://en.wikipedia.org/wiki/Imputation_(statistics)" class="external-link">imputation</a>.
The <code>mice</code> package in R is nice for this.</p>
<p>Note that you do not need to name out every element explicitly. As
long as you provide the objects within each list in the correct order
(data, name, domain, type), you’ll get the correct result:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compactly:</span></span>
<span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_data_list.html">generate_data_list</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">Data1</span>, <span class="st">"genes_1_and_2_exp"</span>, <span class="st">"gene_expression"</span>, <span class="st">"continuous"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">Data2</span>, <span class="st">"genes_1_and_2_meth"</span>, <span class="st">"gene_methylation"</span>, <span class="st">"continuous"</span><span class="op">)</span>,</span>
<span>    uid <span class="op">=</span> <span class="st">"patient_id"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="store-all-the-settings-of-the-desired-snf-runs-in-a-settings_matrix">3. Store all the settings of the desired SNF runs in a
<em>settings_matrix</em><a class="anchor" aria-label="anchor" href="#store-all-the-settings-of-the-desired-snf-runs-in-a-settings_matrix"></a>
</h3>
<p>The <code>settings_matrix</code> is a dataframe where each row
contains all the information required to convert the raw data into a
final cluster solution. By varying the rows in this matrix, we can
access a broader space of possible solutions and hopefully get closer to
something that will be as useful as possible for our context.</p>
<p>In this case, we’re going to create only a single cluster solution
using the same process outlined in the original SNFtool example
above.</p>
<p>An explanation for the parameters in the <code>settings_matrix</code>
can be found at <a href="https://branchlab.github.io/metasnf/articles/settings_matrix.html">the
settings_matrix vignette</a>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">settings_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_settings_matrix.html">generate_settings_matrix</a></span><span class="op">(</span></span>
<span>    <span class="va">data_list</span>,</span>
<span>    nrow <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    alpha_values <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>    k_values <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    t_values <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    dropout_dist <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>    possible_snf_schemes <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">settings_matrix</span></span>
<span><span class="co">#&gt;   row_id alpha  k  t snf_scheme clust_alg cont_dist disc_dist ord_dist cat_dist</span></span>
<span><span class="co">#&gt; 1      1   0.5 20 20          1         2         1         1        1        1</span></span>
<span><span class="co">#&gt;   mix_dist inc_genes_1_and_2_exp inc_genes_1_and_2_meth</span></span>
<span><span class="co">#&gt; 1        1                     1                      1</span></span></code></pre></div>
<p>The columns in this <code>settings_matrix</code> mean the
following:</p>
<ul>
<li>row_id: A way to keep track of the different rows</li>
<li>alpha, k, t: The hyperparameters seen above</li>
<li>snf_scheme: Which “scheme” will be used to transform the inputs into
a final fused network. We’ll discuss this in more detail in the next
vignette.</li>
<li>clust_alg: Which clustering algorithm will be applied to the final
fused network. By default, one of two possible base options are randomly
chosen. A value of 2 indicates that spectral clustering will be used and
that the number of clusters will be determined by the rotation cost
heuristic.</li>
<li>Columns ending in “dist”: Which distance metric should be used. By
default, 1 refers to Euclidean distance for continuous, discrete, and
ordinal data, and 1 refers to Gower’s distance for categorical and mixed
data.</li>
<li>Columns starting with “inc”: Whether or not the corresponding
dataframe will be included for this round of SNF.</li>
</ul>
<p>More detailed descriptions on all of these columns can also be found
in the settings_matrix vignette.</p>
</div>
<div class="section level3">
<h3 id="run-snf">4. Run SNF<a class="anchor" aria-label="anchor" href="#run-snf"></a>
</h3>
<p>The <code>batch_snf</code> function will apply each row of the
<code>settings_matrix</code> (in this case, just one row) to the
<code>data_list</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">solutions_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/batch_snf.html">batch_snf</a></span><span class="op">(</span></span>
<span>    <span class="va">data_list</span>,</span>
<span>    <span class="va">settings_matrix</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Row: 1/1 | Time remaining: 0 seconds"</span></span>
<span><span class="co">#&gt; [1] "Total time taken: 0 seconds."</span></span>
<span></span>
<span><span class="va">solutions_matrix</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span> <span class="co"># it goes on like this for some time...</span></span>
<span><span class="co">#&gt;   row_id alpha  k  t snf_scheme clust_alg cont_dist disc_dist ord_dist cat_dist</span></span>
<span><span class="co">#&gt; 1      1   0.5 20 20          1         2         1         1        1        1</span></span>
<span><span class="co">#&gt;   mix_dist inc_genes_1_and_2_exp inc_genes_1_and_2_meth nclust subject_101</span></span>
<span><span class="co">#&gt; 1        1                     1                      1      2           1</span></span>
<span><span class="co">#&gt;   subject_102 subject_103 subject_104 subject_105 subject_106</span></span>
<span><span class="co">#&gt; 1           1           1           1           1           1</span></span></code></pre></div>
<p>The <code>solutions_matrix</code> is essentially an augmented
<code>settings_matrix</code>, where new columns have been added for each
included patient. On each row, those new columns show what cluster that
patient ended up in.</p>
<p>A friendlier format of the clustering results can be obtained:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster_solution</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_cluster_df.html">get_cluster_df</a></span><span class="op">(</span><span class="va">solutions_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cluster_solution</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cluster  subjectkey</span></span>
<span><span class="co">#&gt; 2       1 subject_101</span></span>
<span><span class="co">#&gt; 3       1 subject_102</span></span>
<span><span class="co">#&gt; 4       1 subject_103</span></span>
<span><span class="co">#&gt; 5       1 subject_104</span></span>
<span><span class="co">#&gt; 6       1 subject_105</span></span>
<span><span class="co">#&gt; 7       1 subject_106</span></span></code></pre></div>
<p>These cluster results are exactly the same as in the original SNF
example:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">cluster_solution</span><span class="op">$</span><span class="st">"cluster"</span>, <span class="va">true_label</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>Running <code>batch_snf</code> with the
<code>return_similarity_matrices</code> parameter set to
<code>TRUE</code> will let us also take a look at the final fused
networks from SNF rather than just the results of applying spectral
clustering to those networks:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">batch_snf_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/batch_snf.html">batch_snf</a></span><span class="op">(</span></span>
<span>    <span class="va">data_list</span>,</span>
<span>    <span class="va">settings_matrix</span>,</span>
<span>    return_similarity_matrices <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Row: 1/1 | Time remaining: 0 seconds"</span></span>
<span><span class="co">#&gt; [1] "Total time taken: 0 seconds."</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">batch_snf_results</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "solutions_matrix"    "similarity_matrices"</span></span>
<span></span>
<span><span class="co"># The solutions_matrix</span></span>
<span><span class="va">solutions_matrix</span> <span class="op">&lt;-</span> <span class="va">batch_snf_results</span><span class="op">$</span><span class="st">"solutions_matrix"</span></span>
<span></span>
<span><span class="co"># The first (and only, in this case) final fused network</span></span>
<span><span class="va">similarity_matrix</span> <span class="op">&lt;-</span> <span class="va">batch_snf_results</span><span class="op">$</span><span class="st">"similarity_matrices"</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>The fused network obtained through this approach is also the same as
the one obtained in the original example:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">similarity_matrix</span> <span class="op">-</span> <span class="va">fused_network</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>And now we’ve completed a basic example of using this package. The
subsequent vignettes provide guidance on how you can leverage the
<code>settings_matrix</code> to access a wide range of clustering
solutions from your data, how you can use other tools in this package to
pick a best solution for your purposes, and how to validate the
generalizability.</p>
<p>Go give the <a href="https://branchlab.github.io/metasnf/articles/a_less_simple_example.html">less
simple example</a> a try!</p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Wang, Bo, Aziz M. Mezlini, Feyyaz Demir, Marc Fiume, Zhuowen Tu,
Michael Brudno, Benjamin Haibe-Kains, and Anna Goldenberg. 2014.
“Similarity Network Fusion for Aggregating Data Types on a Genomic
Scale.” Nature Methods 11 (3): 333–37. <a href="https://doi.org/10.1038/nmeth.2810" class="external-link uri">https://doi.org/10.1038/nmeth.2810</a>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Prashanth Velayudhan, Nicholas Chan, Xiaoqiao Xu, Prajkta Kallurkar, Shihao Ma, Stephanie H Ameis, Ana Patricia Balbon, Adam Taback, Brian Cox, Anna Goldenberg, Bo Wang, Linda Hiraki, Anne L Wheeler, Jasbir Dhaliwal, Colin Hawco, Lauren Erdman.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
